### PR TITLE
Add Subproviders Subpackage

### PR DESCRIPTION
**Pull request by fabioberger**
This PR:
* Creates a new `Subproviders` package with the following subproviders:
    * LedgerSubprovider
    * RedundantRPCSubprovider
    * InjectedWeb3Subprovider
* Adds unit and integration tests for the LedgerSubprovider and unit tests for RedundantRPCSubprovider.
* Use these moved/refactored subproviders in Portal

Open question: We currently also build this package for UMD & CommonJS. Should we only build for CommonJS?

### CODE REVIEW

File: packages/website/package.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
   "author": "Fabio Berger",
   "license": "Apache-2.0",
   "dependencies": {
+    "@0xproject/subproviders": "0.0.0",
```
Comment (LogvinovLeon) [inline packages/website/package.json]: ^0.0.1

File: packages/website/ts/globals.d.ts [hunk 2]
```diff
@@ -157,3 +163,5 @@ declare interface Artifact {
         };
     };
 }
+
+// tslint:enable:max-classes-per-file
```
Comment (LogvinovLeon) [inline packages/website/ts/globals.d.ts]: No need

File: packages/website/ts/globals.d.ts [hunk 3]
```diff
@@ -132,21 +133,26 @@ declare class Subprovider {}
...
+    }
     export = RpcSubprovider;
 }
-// tslint:disable-next-line:max-classes-per-file
-declare class HookedWalletSubprovider {
-    constructor(wallet: any);
-}
-declare module 'web3-provider-engine/subproviders/hooked-wallet' {
-    export = HookedWalletSubprovider;
+declare module 'web3-provider-engine' {
```
Comment (LogvinovLeon) [inline packages/website/ts/globals.d.ts]: We have it in at least two places in code. Maybe it's time for a separate typings package?

File: packages/subproviders/src/subproviders/ledger.ts [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        }
+        const isHexPrefixed = data.slice(0, 2) === '0x';
+        if (!isHexPrefixed) {
+            return false;
+        }
+        const nonPrefixed = data.slice(2);
+        const isValid = nonPrefixed.match(HEX_REGEX);
+        return isValid;
+    }
+    private static validatePersonalMessage(msgParams: PartialTxParams) {
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/subproviders/src/subproviders/ledger.ts]: Maybe in a later iteration.

File: packages/subproviders/src/subproviders/ledger.ts [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+            default:
+                next();
+                return;
+        }
+    }
+    public async getAccountsAsync(): Promise<string[]> {
+        this._ledgerClientIfExists = await this.createLedgerClientAsync();
+
+        const accounts = [];
+        for (let i = 0; i < NUM_ADDRESSES_TO_FETCH; i++) {
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/subproviders/src/subproviders/ledger.ts]: Looks like there is no programmatic way to retrieve the extended public key from a Ledger Nano S of yet.

File: packages/subproviders/src/subproviders/ledger.ts [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+            const result = await this.emitPayloadAsync(payload);
+            this._nonceLock.signal();
+            return result;
+        } catch (err) {
+            this._nonceLock.signal();
+            throw err;
+        }
+    }
+    private async signTransactionWithoutSendingAsync(txParams: PartialTxParams): Promise<ResponseWithTxParams> {
+        await this._nonceLock.wait();
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/subproviders/src/subproviders/ledger.ts]: How so?

File: packages/subproviders/src/types.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    gasPrice?: string;
+    gas: string;
+    to: string;
+    from?: string;
+    value?: string;
+    data?: string;
+    chainId: number; // EIP 155 chainId - mainnet: 1, ropsten: 3
+}
+
+export type DoneCallback = (err?: Error) => void;
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/subproviders/src/types.ts]: It is the way Mocha expects their done callback. Mocha does not follow JS conventions here.

File: packages/website/ts/globals.d.ts (hunk 3 above)
Comment (fabioberger) [reply packages/website/ts/globals.d.ts]: Will move after your PR get merged ;)

File: packages/subproviders/package.json [hunk 8]
```diff
@@ -0,0 +1,53 @@
+{
+  "name": "@0xproject/subproviders",
+  "version": "0.0.1",
+  "main": "lib/src/index.js",
+  "types": "lib/src/index.d.ts",
+  "license": "Apache-2.0",
+  "scripts": {
+    "prebuild": "npm run clean",
```
Comment (BMillman19) [inline packages/subproviders/package.json]: does prebuild get called with `lerna run build` ?

File: packages/subproviders/package.json [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    "clean": "shx rm -rf lib",
+    "build": "tsc",
+    "lint": "tslint --project . 'src/**/*.ts' 'test/**/*.ts'",
+    "run_mocha_unit": "mocha lib/test/unit/**/*_test.js --timeout 10000 --bail --exit",
+    "run_mocha_integration": "mocha lib/test/integration/**/*_test.js --timeout 10000 --bail --exit",
+    "test": "npm run test:unit",
+    "test:all": "run-s test:unit test:integration",
+    "test:unit": "run-s clean build run_mocha_unit",
+    "test:integration": "run-s clean build run_mocha_integration"
+  },
```
Comment (BMillman19) [inline packages/subproviders/package.json]: we need a test:circleci as well

File: packages/subproviders/src/subproviders/ledger.ts [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+        switch (payload.method) {
+            case 'eth_coinbase':
+                try {
+                    accounts = await this.getAccountsAsync();
+                    end(null, accounts[0]);
+                } catch (err) {
+                    end(err);
+                }
+                return;
+
```
Comment (BMillman19) [inline packages/subproviders/src/subproviders/ledger.ts]: should we get rid of all the newlines?

File: packages/subproviders/src/subproviders/subprovider.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+} from '../types';
+/*
+ * A version of the base class Subprovider found in providerEngine
+ * This one has an async/await `emitPayloadAsync` and also defined types.
+ * Altered version of: https://github.com/MetaMask/provider-engine/blob/master/subproviders/subprovider.js
+ */
+export class Subprovider {
+    private engine: any;
+    private currentBlock: any;
+    private static getRandomId() {
```
Comment (BMillman19) [inline packages/subproviders/src/subproviders/subprovider.ts]: some comments here about the approach of producing random Ids might be good

File: packages/subproviders/package.json (hunk 8 above)
Comment (fabioberger) [reply packages/subproviders/package.json]: No, removed this. It's legacy from Dekz's repo.

File: packages/subproviders/package.json (hunk 9 above)
Comment (fabioberger) [reply packages/subproviders/package.json]: Good catch

File: packages/subproviders/src/subproviders/ledger.ts (hunk 10 above)
Comment (fabioberger) [reply packages/subproviders/src/subproviders/ledger.ts]: No, our convention is to keep new lines between switch cases **if** the cases are not wrapped in curly-brackets. I'm happy to revise this, but I feel it better elucidates the separation between the cases.

File: packages/subproviders/src/subproviders/ledger.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/subproviders/src/subproviders/ledger.ts]: Are you sure? https://github.com/LedgerHQ/ledger-node-js-api/blob/e7b12b1911f265c3e782bc5bf6d890aacfe74b2a/src/ledger-eth.js#L27

File: packages/subproviders/src/subproviders/ledger.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/subproviders/src/subproviders/ledger.ts]: We can lock inside of `populateMissingTxParamsAsync`

**Pull request merged**